ddl/column.go

```diff
@@ -427,13 +427,118 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col := findCol(tblInfo.Columns, newCol.Name.L)
+	if col == nil {
+		col = findCol(tblInfo.Columns, oldColName.L)
+	}
+
+	originalState := col.State
+	switch col.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, newCol, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
+		// otherwise `insert into t value (1, 2)` will add different row and index record.
+		newCol.State = model.StateDeleteOnly
```

Comment:
Now we not change the column's state. So it will allow `insert t value(1, 2)` execute as `insert t(a, b) value (1, 2)` or `insert t(b, a) value (1, 2)` before the `modify column` ddl job done. It seems OK to allow this.